Sources/Workspace/Workspace.swift

```diff
@@ -2554,9 +2562,31 @@ public final class LoadableResult<Value> {
     }
 }
 
+private struct RemoteArtifact {
+    let packageRef: PackageReference
+    let targetName: String
+    let url: String
+    let checksum: String
+}
+
+private extension ManagedArtifact {
+    var originURL: String? {
+        switch self.source {
+        case .remote(let url, _):
+            return url
+        case .local:
+            return nil
+        }
+    }
+
+    func kind() throws -> BinaryTarget.Kind {
+        return try BinaryTarget.Kind.forFileExtension(self.path.extension ?? "unknown")
```

Comment:
Should this throw if there's no extension, or is it useful to support unknowns?

Reply:
its a bit subtle but it will throw on "unknown", since there is not such extension